okio/jvm/src/main/java/okio/PeekSource.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package okio
+
+/**
+ * A [Source] which peeks into an upstream [BufferedSource] and allows reading and expanding of the
+ * buffered data without consuming it. Does this by requesting additional data from the upstream
+ * source if needed and copying out of the internal buffer of the upstream source if possible.
+ *
+ * This source also maintains a snapshot of the starting location of the upstream buffer which it
+ * validates against on every read. If the upstream buffer is read from, this source will become
+ * invalid and throw [IllegalStateException] on any future reads.
+ */
+internal class PeekSource(
+  private val upstream: BufferedSource
+) : Source {
+  private val buffer = upstream.buffer
+  private var expectedSegment = buffer.head
+  private var expectedPos = buffer.head?.pos ?: -1
+
+  private var pos = 0L
+
+  override fun read(sink: Buffer, byteCount: Long): Long {
+    if (expectedSegment != null) {
+      // Source becomes invalid if there is an expected Segment and it and the expected position
+      // do not match the current head and head position of the upstream buffer
+      check(expectedSegment == buffer.head && expectedPos == (buffer.head?.pos ?: -1)) { "invalid" }
```

Comment:
nit: `===` for reference equality on segment comparison (not that it changes behavior, you're going to do this anyway but this gets there faster)